### ISSUE TITLE
[ENH] umbrella issue - categorical feature support

### ISSUE THREAD

**Issue by fkiraly** [sktime/sktime#6109]
Issue to collect material and input on adding categorical feature support, and summarizing discussion from various locations. For discussion.

### Aim

It has been requested a number of times to add support for categorical features in classification/regression (features) and forecasting (endogenous and exogenous data).

### Basic requirements

What does this mean:

* classifiers/regressors and forecasters *and transformations* can in-principle accept categorical features
    * not all estimators will support this out of the box, so I anticipate a tag is needed to distinguish those that can, from those that cannot
    * this may be complicated by `sklearn` transformations that are not fully inspectable as to this property (afaik)
* major compositions such as pipelines, reduction, ensembles also deal with categorical variables and the new tags correctly
* at least two estimators of each type are implemented, for testing and demonstration
    * and there are worked examples with composites and transformations
* tests are added for the above

### architectural issues, open design questions

* currently, checks and conversions (localized in the `datatypes` module) do not allow categorical variables through. This is because `object` `dtype` in time series containers was no tallowed, and not all time series containers support categorical type
    * PR https://github.com/sktime/sktime/pull/5886 allows object `dtype` for the time series containers.
    * open design question: it is not entirely clear how the distinction between containers that can represent categorical data (or mixed categorical/numerical) and those that cannot should be handled. Noting that 0/1 encoding in `numpy` must be treated as numerical since there is no intrinsic distinction between the two, and no extrinsic coding is available
    * open design question: how should conversions between data types be handled, between those that can or cannot? E.g., always apply a one-hot encoding if converting to `numpy`? This might substantially alter the logic in the converters as current. Or, assume this case never occurs?
    * related refactor: https://github.com/sktime/sktime/pull/6033, this could carry further tags for datatypes with "categorical capability", similar to estimators?
* tests need to be expanded to handle categorical variables. This should at least include systematic, scenario based testing with categoricals: https://github.com/sktime/sktime/pull/5964
    * one would expect that the test should check that the right error message is raised if the "categorical" tag is false, otherwise that all tests run through.

### possible work plan

Some thoughts about work plan:

* I think this is sufficiently complex that we need an enhancement proposal written up, for the desired end state, example code, a design that solves the above issues, and a list of estimators (native and/or interfaced, preferably one example of each per learning task) to implement or extend
    * this should also include a detailed implementation plan
* ths first implementation item is likely the data container framework, with an interplay of tags, container specifications, converters. The sequencing is not entirely clear, I would start with the existing PR and first add the tag (saying no estimator can support categorical at first), then slowly switch on support while further iterations on the framework take place.
* based on this, I'd add a few examples of forecasters, classifiers, transformations - preferably simple ones
* next are pipelines, compositions. Reduction is important, though it is possibly the most unpleasant to extend.

**Comment by fkiraly** [comment]
FYI @yarnabrina, @Abhay-Lejith

**Comment by yarnabrina** [comment]
> open design question: in forecasters, do we need to distinguish categoricals in endogeneous vs exogeneous data? E.g., by tag, or by how we handle them?

This is the one that's most interesting and confusing to me at the moment. Are there any forecasting algorithm (in or out sktime) which actually supports categorical variables (without encoding of any type)?

When I created the issue my thouwas entirely about exogenous features only. But if someone uses that with ForecastX like compositions, this is something we have to figure out. So I think the answer is _yes_ to the question as future known X does not need a specific algorithm while y or future unknown X will need.

**Comment by nanamikato** [comment]
@fkiraly If I'm not wrong, you have recently created an enhancement proposal for the project "time-to-event modelling, probabilistic forecasting, probabilistic supervised learning" similar to the one above for categorical feature support, correct? Can I follow that and utilize those enhancement proposals to develop my gsoc proposal?

**Comment by nanamikato** [comment]
@fkiraly If I'm not wrong, you have recently created an enhancement proposal for the project "time-to-event modelling, probabilistic forecasting, probabilistic supervised learning" similar to the one above for categorical feature support, correct? Can I follow that and utilize those enhancement proposals to develop my gsoc proposal?

**Comment by fkiraly** [comment]
> Are there any forecasting algorithm (present or absent in sktime as of now) which actually supports categorical variables (without encoding of any type)?

This is off-topic for this issue, but yes. Kindly note, I just merged them as they have been open for a year and it appears we forgot to merge. Many of the features in the proposal are already implemented, or specifications have drifted.

**Comment by yarnabrina** [comment]
@fkiraly can't say I follow your second comment above. Is it by any chance for some different thread?

**Comment by fkiraly** [comment]
I think there was a comment by someone named nanami, or similar, inquiring about the best starting point in the probabilistic forecasting project. The comment was along the lines, whetherh the STEP I recently merged in the enhancement-proposals repository was a good starting point. I am fairly sure it was in this issue, as I also note that it is off-topic. I assume the poster deleted the message, e.g., after noticing that it was off-topic. Possibly they meant to post elsewhere.

In case you're reading this, that's no problem, please do not be worried! It's fine to post off-topic.

**Comment by Abhay-Lejith** [comment]
One question I have is, are we not going to implement/interface an encoder to convert categorical values to numerical, which can be used by users to convert their data? Or are we going to do the conversion internally itself? 
Because in the first case, we wouldn't need tags to specify which estimators can handle categorical data since users will themselves encode the data and use the estimator normally.
Or is the tag to distinguish b/w estimators that can handle categorical data on their own without encoding and those that cannot?

**Comment by fkiraly** [comment]
> One question I have is, are we not going to implement/interface an encoder to convert categorical values to numerical, which can be used by users to convert their data? Or are we going to do the conversion internally itself?

I would (a) leave that to user choice, since there are multiple ways to encode, and (b) ensure that we are compatible with the `sklearn` encoders.

There is an issue around the deisgn where these transformations accept only `y`, not `X`, see here: https://github.com/sktime/sktime/issues/5867. They also do not seem to have the "natural" inverse transform, decoding.

**Comment by Abhay-Lejith** [comment]
I didn't see this mentioned anywhere in the prior discussions/PRs but `sklearn` also offers an `OrdinalEncoder` which performs the same function as LabelEncoder but only on `X `and not `y`. And both encoders have inverse transform methods which perform decoding back to original values. Do we want something which accepts both `X` and `y`? Currently nothing of that sort is present in sklearn so we would have to probably do something by making use of both these encoders or making our own.

**Comment by fkiraly** [comment]
@Abhay-Lejith, I think the `sklearn` design to have the same estimator basically with two different names for `fit` argument is bad design. In-principle, this would lead to having to duplicate every transformer, once with `X` and once with `y` argument - see discussion in #5867.

I've tried to solve this with an update to the `sklearn` adapter which maps everything on the `X` argument, and whether it is applied to `X` or `y` depends on how you specify the pipeline, not on the transformer itself, see https://github.com/sktime/sktime/pull/5982

**Comment by Abhay-Lejith** [comment]
I think we could begin with this as you had suggested @fkiraly, by introducing a tag to differentiate between estimators that can and cannot handle categorical variables. Some of the tests in #5964 are failing due to passing categorical variables to non-compatible estimators. We can then check the value of this tag and return proper error messages if categorical variables are passed to an estimator whose tag indicates that it cannot handle it. And I think for those that can handle categorical variables, they will still need some parameters changed accordingly which I guess will have to be done depending on the nature of the input.

> open design question: in forecasters, do we need to distinguish categoricals in endogeneous vs exogeneous data? E.g., by tag, or by how we handle them?

I am not very sure about what would be best here.

**Comment by Abhay-Lejith** [comment]
Also, since necessary changes have already been made with respect to modifying the tabular to series adapter as well as allowing categorical variables in checks, the next move should be to test if these changes work right? Or are there any other components that also need to be taken care of?
I'm just trying to get as much information as possible so I can come up with a good project timeline for my proposal.

**Comment by fkiraly** [comment]
> I think we could begin with this as you had suggested @fkiraly, by introducing a tag to differentiate between estimators that can and cannot handle categorical variables. Some of the tests in https://github.com/sktime/sktime/pull/5964 are failing due to passing categorical variables to non-compatible estimators. We can then check the value of this tag and return proper error messages if categorical variables are passed to an estimator whose tag indicates that it cannot handle it. And I think for those that can handle categorical variables, they will still need some parameters changed accordingly which I guess will have to be done depending on the nature of the input.

Yes, I think that would be a good start!

**Comment by fkiraly** [comment]
> Also, since necessary changes have already been made with respect to modifying the tabular to series adapter as well as allowing categorical variables in checks, the next move should be to test if these changes work right? Or are there any other components that also need to be taken care of?

Good question. I think it is important to come up with a few good test cases of pipelines and similar, and develop against these.

**Comment by Abhay-Lejith** [comment]
Which estimators in sktime support categorical values out of the box? I could only find a few and those were in the classification deep learning estimators. ( they do the encoding themselves internally )
And as far as I understand, estimators that do not support it themselves cannot simply be 'extended' to support categorical variables. Encoding and then inputting the data is the only option.

>and a list of estimators (native and/or interfaced, preferably one example of each per learning task) to implement or extend

So I didn't quite understand what exactly you meant here @fkiraly. I also cannot seem to find any forecasting algorithms that use categorical data directly as such. I'm confused about what can be implemented/extended.

**Comment by fkiraly** [comment]
> Which estimators in sktime support categorical values out of the box?

I think this is the same question, please let me know if I misunderstand, that @yarnabrina has aked above, there is a list:

https://github.com/sktime/sktime/issues/6109#issuecomment-1989464887

**Comment by fkiraly** [comment]
>> and a list of estimators (native and/or interfaced, preferably one example of each per learning task) to implement or extend

> So I didn't quite understand what exactly you meant here @fkiraly.

I meant, we should prioritize new ones to implement or interface - as opposed to existing estimators which can gain categorical support by minor adaptation.

**Comment by Abhay-Lejith** [comment]
> I think this is the same question, please let me know if I misunderstand, that @yarnabrina has aked above, there is a list:
> 
> [#6109 (comment)](https://github.com/sktime/sktime/issues/6109#issuecomment-1989464887)

Yes I meant the same thing but I was wondering if you could list their names as it is in sktime because I was unable to find their codes in the repo like for example, the hidden markov models.

**Comment by Abhay-Lejith** [comment]
>And as far as I understand, estimators that do not support it themselves cannot simply be 'extended' to support categorical variables. Encoding and then inputting the data is the only option.

Also clarification on this would be greatly appreciated. I say this because the algorithms are made to deal with numeric variables performing various complex operations with them. What slight modifications could possibly make it compatible with categorical variables? In every ML task I've come across so far, encoding is the only option I've seen.

**Comment by fkiraly** [comment]
> Yes I meant the same thing but I was wondering if you could list their names as it is in sktime because I was unable to find their codes in the repo like for example, the hidden markov models you mentioned.

* `DirectTabularRegressionForecaster`, `DirectReductionForecaster`
* `NaiveForecaster`
* `LTSFLinearForecaster`, for example (for DL)
* `GMMHMM`, although this is unsupervised, `hmmlearn` can be used for supervised mode too

**Comment by duydl** [comment]
I would like to mention that this issue is possibly also related to the work on annotation module design i.e #3214. 

All currently implemented annotation algorithms are unsupervised, with an interface analogous to the transformation module. So I think it is necessary to think about the design for a different category namely supervised tasks, which I assumed contains time-series with labels for points/segments. That means the design for an annotation base class would be related to how transformation could be extended to support the labels series as an additional categorical feature.

**Comment by fkiraly** [comment]
I see! Very interesting point!!

Let me check whether I understand this right, please confirm:

* you are saying categorical feature support is relevant to the annotation base class (re-)design, because time series annotations are semantically/mathematcially equivalent to a categorical time series.
* Further, if we continue in exogenous/endogenous terminology, in both settings we would have free-type `X`, and categorical `y`.
* Even further, forecasting with univariate endogenous categorical series and future-known exogenous data has the same abstract type as supervised stream/online annotation (fit on `X` and categorical `y`, predict `y` from, and at indices of `X`).

?

**Comment by duydl** [comment]
I could confirm that the first two points are what I meant. For the third point, I am not yet sure how much stream processing would be related to forecasting.

**Comment by fkiraly** [comment]
> we could consider that the algo generates its own exogenous y from the initial window.

That sounds like some kind of reduction? I'd locate this with reduction at least, not on the level of interface.

**Comment by Abhay-Lejith** [comment]
Ok so I was just looking at the code for `NaiveForecaster` and trying it out with categorical variables in a branch checked out from the pr ( #5964 ) with the modifications to allow object dtype. Only the `last` strategy can be made to work with slight modifications. Even other simple strategies like `mean` and `drift` do not work as they involve steps like subtracting values to find the slope and dividing values. How can it be extended to support categorical values? Other more complex models will surely have more mathematical operations being performed on the data. This is what I am trying to ask @fkiraly. I feel like I am not understanding something here.

**Comment by fkiraly** [comment]
> Other more complex models will surely have more mathematical operations being performed on the data. 

Yes, not all `strategy` settings will support categorical, so it might be interesting to think about what to do here. That is, sth like `capability:categorical` depending on parameter values?

For instance, I could think of some strategies from `DummyClassifier` being used here: https://scikit-learn.org/stable/modules/generated/sklearn.dummy.DummyClassifier.html, like "most frequent".

I'm not trying to prescribe here, just pointing to some open design questions. There are various design issues to resolve, e.g., should there be different estimators, or just some strategies only work depending on the type ot `y` (categorical or not).

**Comment by fkiraly** [comment]
> I feel like I am not understanding something here.

I think your points are very valid, as you are trying to make the design very concrete. This is very important work that has now highlighted the important questions above. The next step would be thinking about alternative solutions and directions, pros/cons.

**Comment by Abhay-Lejith** [comment]
> take any DL forecasting model and switch the endogenous loss to categorical

My understanding is that simply changing the loss would not allow the model to accept and work with categorical variables. It is just the loss being calculated in a different manner. The internals are still doing a lot of math which cannot be done on object dtyped variables. (and `LSTFLinearForecaster` does not have categorical loss functions either). 

This is essentially my thought:
I dont think any model would work out of the box with categorical variables.

The `NaiveForecaster` on `last` strategy would probably be the `only exception` to this because it does nothing mathematical to the data , just repeats the last value from fit.

So what I'm trying to say is that there will be no model that can handle categorical values without encoding. Hence the idea of extending models to support categorical features or implementing/interfacing new ones (none such exist as far as I could search on the internet) is not possible.

**Comment by fkiraly** [comment]
> and `LSTFLinearForecaster` does not have categorical loss functions either

not sure if I get that - can you not change `criterion` to any `torch.nn` loss, including, say, `CrossEntropyLoss`?

**Comment by fkiraly** [comment]
> I tried adding `CrossEntropyLoss` but that throws up errors due to categorical variables from the internals of `pytorch` dataloaders.

Sure, but "making it work" would ultimately just involve adding some array type coercion in this case, no? That's a "minor change" for me.

**Comment by fkiraly** [comment]
> This is of course in much less scope and much less detailed that what @fkiraly described above, but just sharing for reference (mainly for my own benefit to check later if these expectations as user are fulfilled or not).

@yarnabrina, I do agree that we need to start with a modest scope and low hanging fruit, at least to proof-of-concept the framework extensions.

**Comment by Abhay-Lejith** [comment]
@fkiraly the point I'm trying to get across is, all models perform some mathematical operations on input data to finally make predictions. No change to the code is gonna allow these operations to work on non-numeric data. So changing the loss, or performing some array type coercion is not going to change the fact that the model will throw an error when it tries to perform some mathematical operation on non-numeric data.

**Comment by fkiraly** [comment]
> No change to the code is gonna allow these operations to work on non-numeric data. So changing the loss, or performing some array type coercion is not going to change the fact that the model will throw an error when it tries to perform some mathematical operation on non-numeric data.

Not sure if I follow along - type coercion, conversion and/or encoding will, if it is done "under the hood". For instance, in case of deep learning, if you do a minor case distinction and convert the arrays to numeric suitably, that should cover the categorical case. I would consider this a minor change, since the modification to the logic is not of substance.

**Comment by Abhay-Lejith** [comment]
I think there is a misunderstanding here, I though you had said we will not be doing any such conversion/encoding 'under the hood' and will leave it to the user to encode their categorical features earlier here https://github.com/sktime/sktime/issues/6109#issuecomment-1995431336.

**Comment by fkiraly** [comment]
That was about the general boilerplate - in individual estimators, it may make sense. There's also a difference between full-blown encoding and dtype coercion.

**Comment by Abhay-Lejith** [comment]
I see, could you please elaborate on what exact type of coercion could be used here? Like how it could be used to convert to numeric data.

**Comment by fkiraly** [comment]
> I see, could you please elaborate on what exact type of coercion could be used here? Like how it could be used to convert to numeric data.

It depends - the simplest case might be just a coercion of a two-class categorical to float 0/1, or multi-class categorical to integer.

This can be considered a kind of encoding, but it's different from, e.g., one-hot-encoding, where you get a 0/1 numerical matrix (n_samples, n_classes) in the end.

**Comment by Abhay-Lejith** [comment]
I see, so you mean to say a kind of encoding where we don't generate extra columns essentially and instead just replace the categorical variables with numeric values in the same column itself. But wouldn't something like that create an ordering of sorts among the categorical variables which would be undesirable?

**Comment by fkiraly** [comment]
> But wouldn't something like that create an ordering of sorts among the categorical variables which would be undesirable?

yes, "in general" that is not a good idea, but for specific estimators it might work out fine - depending on what they do internally.

**Comment by Abhay-Lejith** [comment]
Ok so I had suggested the idea of adding parameters like `enable_categorical` to estimators for which support is extended but you said that might not be a good idea. Could you please explain why?

The other option I see is, we allow the user to input any data without specifying such parameters and then check for object dtype. And then depending on the value of the 'support categorical' tag of the estimator, we throw an error or move to converting the categorical to numeric data. Is this the better option?

**Comment by Abhay-Lejith** [comment]
What I had in mind was that the `enable_categorical` tag would be set by the user depending on whether his input contains categorical values or not. Depending on this we could decide whether to perform the dtype conversion or not. 

pros:
- it enables us to distinguish between whether categorical X or y has been passed. ( using `enable_categorical_X` and `enable_categorical_y` ) . More on why this is important at the end.
cons:
- we might have to add the parameter to a lot of, if not all estimators? Coming to this below.

>we would have to add this parameter to every estimator that eventually will support categoricals (or, every estimator?).

About this, if we are going to simply add this categorical support to all/most estimators despite the fact that label encoding creates an ordering among variables which is undesirable and would reduce the effectiveness of the estimators predictions, then what would be the point of providing a separate encoder at all. I was thinking we'll extend support only to a few simple models.

>The other option I see is, we allow the user to input any data without specifying such parameters and then check for object dtype. And then depending on the value of the 'support categorical' tag of the estimator, we throw an error or move to converting the categorical to numeric data.

pros:
- No extra parameters added, less major design changes.

cons:
- Cannot distinguish between whether categorical X or y has been passed. Maybe there is some other way to make the distinction?

Coming to why I think making this distinction is important is, if categorical `y` is passed, we will be required to perform an inverse transform on the final result ( predict after fitting ) before displaying it to the user. In order to know whether this inv transform needs to be done or not, it would help to know whether categorical `y` was passed to fit in the first place. The `enable_categorical_y` parameter would aid this.

**Comment by fkiraly** [comment]
>What I had in mind was that the `enable_categorical` tag would be set by the user depending on whether his input contains categorical values or not. Depending on this we could decide whether to perform the dtype conversion or not.

Do you mean tag or parameter?

Either way it would be passing redundant information, since isn't it implied by the data container, whether categorical variables are present or not?

> Cannot distinguish between whether categorical X or y has been passed. Maybe there is some other way to make the distinction?

Same answer. Is that not directly inspectable in the data container?

**Comment by Abhay-Lejith** [comment]
>Do you mean tag or parameter?

Sorry, I meant the parameter there.

>Same answer. Is that not directly inspectable in the data container?

Is it possible to examine the `y` passed to `fit` when the `predict` method is called?

**Comment by yarnabrina** [comment]
> Is it possible to examine the y passed to fit when the predict method is called?

Yes. It's stored in `self._y`.

> if we are going to simply add this categorical support to all/most estimators despite the fact that label encoding creates an ordering among variables which is undesirable and would reduce the effectiveness of the estimators predictions, then what is the point of even doing it

While this makes sense, I think this choice should be left to sktime users whether they want to this encoding or not. We can surely throw an warning noting concerns, but failing outright seems very restrictive.

Also we do not necessarily have to stick to label encoding. In some cases, user may choose to go for ordinal encoding themselves with specified choices of category orders. I haven't fully followed all the discussion above, but have we ruled out all other forms of encoding already, e.g. the super popular one hot?

> Coming to why I think making this distinction is important is, if categorical y is passed, we will be required to perform an inverse transform on the final result ( predict after fitting ) before displaying it to the user. In order to know whether this inv transform needs to be done or not, it would help to know whether categorical y was passed to fit in the first place. The enable_categorical_y parameter would aid this.

I think this is a genuine scope question. @fkiraly I was thinking that this work is primarily going to focus on categorical support for X features, assuming we are sticking to forecasting for this discussion. If categorical y is also being considered, then I think encoding may affect the result significantly for models that don't support it natively. The concerns raised by @Abhay-Lejith above regarding ordering and etc. is very important. Plus we need to somehow restrict the predictions to be integers, otherwise we are assuming equidistant categories along withe already assumed ordering.

---

Also, have we considered distinction between "object" and "category" dtypes? Will that matter anywhere? If it is present with ordering, that may give some additional info.

**Comment by fkiraly** [comment]
I see, @yarnabrina, @Abhay-Lejith, valid points.

From the discussion, I would conclude that scope definition should be at the very start.
I would suggest to do this with code snippets that we want to work in the first phase of this, and proceed to an STEP (extension proposal = design document) from there.

I suppose pipelining with `sklearn` one-hot-encoder working end-to-end is the most basic requirement? Or, what do you think, @yarnabrina?

**Comment by Abhay-Lejith** [comment]
Yes please, if something like that could be done then I can come up with a concrete plan for my proposal. At the current stage it is difficult for me to make an implementation plan with no specifics decided on. The deadline for submission is just a week away too.

**Comment by yarnabrina** [comment]
I'll share a draft version by tomorrow night IST. It's unlikely to be very formal or very detailed, but will list down what are the things I expect to be solved as initial focus from users' point of view.

**Comment by yarnabrina** [comment]
Here's my attempt to pen down my vision for this project, at least to start with.

---

# Top Level Goals

1. As a user, I want to be able to pass a dataset with `sktime` which has non-numeric columns.
    - The objective is if someone uses `sktime` only as a middle layer for transformations or something similar, they should be blocked because we completely disallow non-numeric columns. Users should be able to use any transformations at least with `ColumnSelect`.
    - As part of this expectation, I think the idea is that restrictions should come from the estimators themselves instead of being imposed the framework itself.
    - I think PR #5886 is probably the start of this expectation. @fkiraly may guide on areas where it is not sufficient.
    - Fulfilling just this expectation is insufficient to generate informative messages for users. This will be addressed in point 2 below.
    - Fulfilling just this expectation is certain to cause failures from a lot (if not all) of estimators who do not support non-numeric features at all. This will be addressed in point 2 below.
2. As a user, I want to be aware of the reason of failures in a user friendly fashion.
    - The objective is to let the user know if the error is generated out of code issues or specifically because of inability of estimators to deal with non-numeric variables.
    - As part of this, I think the idea should be to not pass non-numerical data to an estimator incapable of non-numeric data and fail with a specific error class and/or message.
    - I think the way to address this will be to add tag to all existing concrete estimators (directly or indirectly through inheritance) in `sktime`. For `sklearn` estimators, we can attempt to use the `X_types` tag provided by `sklearn` and check for `categorical` being present or not.
3. As a user, I want be able to control whether to encode non-numeric variables or not.
    - The objective is to not encode automatically without giving users an option to control that behaviour.
    - As part of this, I think the idea is to have an option set on an estimator (or may be on a pipeline) to allow encoding or not. If not allowed, failures for non-numeric data is expected by users with informative messages. Otherwise, `sktime` should (try to) encode it in a way to ensure no error due to non-numeric data.
    - I think the way to address this will some sort of configuration option that users can select with `set_config` or similar.
    - This expectation does not describe the process of encoding to perform. This will be addressed in point 5.
4. As a user, I want to be able to control the method to encode non-numeric data.
    - The objective is to not encode automatically with a fixed method without giving users a choice.
    - As part of this, I think the idea is to have an option set on an estimator (or may be on a pipeline) to control encoding method. If not provided, `sktime` should refuse to accept the instruction to encode before passing to estimators without support for non-numeric data.
    - I think the way to address this will some sort of configuration option that users can select with `set_config` or similar.
    - This expectation does not describe the process of encoding to perform. This will be addressed in point 5.
5. (Not a user goal) Define or allow definition of encoding options
    - `LabelEncoder` (or something equivalent, e.g. `pandas.Categorical.codes`) is an obvious option. It is the simplest and most intuitive, however it adds assumptions of ordering and equidistant categories.
        - I believe `OrdinalEncoder` practically behaves identically (except the signature), that can be considered together in same group.
    - `OneHotEncoder` is another widely used option. It does not assume order or equidistance, but can create too many columns with high cardinality variables. It also can lead to non-full rank scenario and can cause issues with respect to matrix invertibility.
    - In context of classification or regression, `TargetEncoder` is quite popular, especially in case of high cardinality variables.
    - [`category-encoders` library](https://pypi.org/project/category-encoders/) provide a lot of other encoders, which can be considered.

## Notes

- I am using non-numeric term consistently in case we decide to distinguish between object/category/string dtypes later.
- I am suggesting these points as expectations from a `sktime` user, primarily of forecasting component.
- I have deliberately kept non-numeric y out of these. Specifically for classification component, this is likely to be an issue, may be @fkiray can extend these?
- I have consistently assumed that encoding should never modify the data permanently and will only be done to pass non-numeric data to an estimator that does not have such support. If there is a pipeline where many (or all) estimators are like that, this may lead to encoding too many times. An option will be then to encode only once for a sequence of estimators that lacks the ability, and may be @benHeid can suggest on how to detect such either in sequential or in graphical pipeline?
- The above note applies only for encoding non-numeric data when `sktime` is configured to do as such as per points 3 and 4 above. No modification expectation does not apply in case user creates an explicit `sklearn` encoders through `TabularToSeriesAdaptor`.

**Comment by fkiraly** [comment]
(fixed a few typos, I felt these were obvious typos. Please correct me if wrong)

**Comment by fkiraly** [comment]
FYI @yarnabrina, agree with the high-level goals. Not sure if configs are the best way to address points 3 or 4. My thinking is, estimator logic should be kept out of config fields, and should always be accessible through the parametric specs (nested class constructors).

**Comment by yarnabrina** [comment]
I am also not sure if set_config is the best way, but changing signature by adding 2 new keyword arguments at end for all estimators seemed too big (and potentially risky). I am not sure if that's what you meant by nested constructors, if not can you please elaborate?

**Comment by fkiraly** [comment]
@yarnabrina, why would we change signatures?

I'm also opposed to adding n kwargs to all estimators, if this can be avoided.

> I am not sure if that's what you meant by nested constructors, if not can you please elaborate?

What I'm trying to say, the math/formal logic should be completely determined by expressions like

`Foo(bar=42, MyEstimator("a", "b", c="c")) * Something("else")`

If `set_config` controls the way we encode, that would violate this principle.

**Comment by yarnabrina** [comment]
> @yarnabrina, why would we change signatures?

I thought you meant change `Est(a=1, b=2)` to `Est(a=1, b=2, convert_categorical=True, conver_categorical_method="label")` something like that, so that's why I said that. Apparently that's not the case.

`set_config` was just an option that I thought of, I have absolutely no strong preference at all for that. Any suggestion for points 3/4 is totally fine by me.

**Comment by fkiraly** [comment]
> I thought you meant change `Est(a=1, b=2)` to `Est(a=1, b=2, convert_categorical=True, conver_categorical_method="label")` something like that, so that's why I said that. Apparently that's not the case.

No, I wasn't advocating that.

In addition, if there are parameters like this, it's better to have composition, e.g., `LabelEncoder() * Est(a=1, b=2)`, no?

**Comment by fkiraly** [comment]
> `set_config` was just an option that I thought of, I have absolutely no strong preference at all for that. Any suggestion for points 3/4 is totally fine by me.

I think it's not an unreasonable option, it just departs from the `sklearn` paradigm a bit.

**Comment by Abhay-Lejith** [comment]
@fkiraly just to make it very clear, could you give an example code snippet of how you expect the equivalent of this ( `Est(a=1, b=2, convert_categorical=True, conver_categorical_method="label")` ) to look like?
because I didn't quite get what you meant by this
>and should always be accessible through the parametric specs (nested class constructors).

**Comment by fkiraly** [comment]
Sure: `LabelEncoder() * Est(a=1, b=2)` - for example

**Comment by yarnabrina** [comment]
While I don't have any strong objections, I'll note that this is not directly in line with my goals and notes above. That's absolutely fine, but I'll note the two deviations below in case it may have some other impacts.

1. It's basically combining points 3 and 4 above and contrasting second last bullet in the notes.

2. It is not really an encoding mechanism provided by sktime to allow indirect categorical support to estimators that lacks such, it's just forcing users to add a new estimator in sequence if they want categorical. While in a lot of situations that's okay and may be actually better, it probably doesn't allow users to configure different types of encoding per different variables or per different estimators in a sequence if it's known that one estimator is especially bad with a particular kind of encoding.

3. This is going to encode and modify the variable always, and won't pass actual categorical values to an estimator in a pipeline which supports that natively.

**Comment by Abhay-Lejith** [comment]
I think what @fkiraly is suggesting is different from what @yarnabrina is proposing. 
The example given by @fkiraly is to use a `LabelEncoder` explicitly outside of the estimator to encode values.
On the other hand I think @yarnabrina is suggesting to do this encoding internally itself within the estimator with the user passing whether the encoding must be done and by what method through some parameters or set_config or some other way if possible.

There is still confusion regarding what options we want to give a user to encode their variables. 
>Sure: LabelEncoder() * Est(a=1, b=2) - for example

This example for instance doesn't involve any changes in the estimator as the encoder simply does its thing and pipes the input with the encoded values into the estimator. Please correct me if I've understood something wrong.

**Comment by Abhay-Lejith** [comment]
@fkiraly the example you've given above is confusing me. We were discussing how to let the user specify whether to encode or not and the method of encoding in the estimator through methods such as extra parameters or set_config. The example you've mentioned above is completely different and is using a separate LabelEncoder instance.

**Comment by fkiraly** [comment]
@Abhay-Lejith, apologies from being unclear.

My comments aim to highlight, that there are two layers to the problem, which imo are distinct but related:

1. ensuring that the framework supports estimators and pipelines with categorical features, including categorical encoders
2. implementing categorical encoding in specific estimators

So, we need to make sure that sth like `LabelEncoder() * Est(a=1, b=2) ` would *always* work, if `Est` supports numerical data. This is point 1.

Now, `Est` could also decide to do some automatic encoding internally, if it encounters categorical data. Similar to how already now some classifiers apply internal one-hot encoding, e.g., see discussion in https://github.com/sktime/sktime/issues/5864.

In this case, the encoding would be the responsibility of `Est`.

My point is, it is far more important to ensure requirement 1, because amongst others it is a pre-condition that 2 is even possible. Further, it is not desirable to do number 2 for all estimators.

Please let me know if sth is unclear or you have questions (or if you disagree).

**Comment by Abhay-Lejith** [comment]
> 2. implementing categorical encoding in specific estimators
@fkiraly, What is the point of this? When the option to encode using encoders exists, why would we want to simply add this automatic internal encoding feature to some estimators randomly?

**Comment by fkiraly** [comment]
> When the option to encode using encoders exists, why would we want to simply add this automatic internal encoding feature to some estimators randomly?

Not randomly - in estimators where it is "straightforward" or "canonical". See, e.g., what @yarnabrina said here: https://github.com/sktime/sktime/issues/6109#issuecomment-2007509387

**Comment by Abhay-Lejith** [comment]
The estimators referred to there are those which already have inbuilt encoding schemes. We do not have to make any modifications to those estimators. 
>in estimators where it is "straightforward" or "canonical"

I'm not sure if I follow, in any estimator, adding support would simply involve encoding the categorical variables. It is the same procedure for any estimator.

**Comment by fkiraly** [comment]
> I'm not sure if I follow, in any estimator, adding support would simply involve encoding the categorical variables. It is the same procedure for any estimator.

I think that's incorrect?

For example, to enable support in direct reduction (`DirectReducer`) for categorical endogenous `y`, you would have to allow the estimator to be a classifier. That is different, mathematically and logically, from encoding `y`.

**Comment by Abhay-Lejith** [comment]
I see, I wasn't thinking of this example. I was thinking with respect to regular forecasters such as ARIMA, autoETS, SARIMAX, STLforecaster ,ThetaForecaster etc...

**Comment by fkiraly** [comment]
> Also, which sklearn classifier accepts categorical variables ?

If we're talking about the target vector - which is what "endogenous" translates to when used in forecasting - then, all of them.

**Comment by fkiraly** [comment]
By tangent I got aware of a barely visible (?) project of Rob Hyndman, which extends the `fable` package (like `sktime` forecasting module, but in R) for categorical, binary, endogenous variables: https://github.com/tidyverts/fable.binary

The pathway here is @ericjb's PR #6416, which leads to `tsibbles` (like `pd-multiindex` but its own data container and in R), which leads to `tidyverts` (Hyndman GitHub org), which contains the project.

FYI @Abhay-Lejith, @yarnabrina

**Comment by fkiraly** [comment]
Some observations:

* `skforecast` does not "internally" support categorical features, it requires the user to build pipelines with encoders etc. However, there are a few recommended patterns that one could interface instead, leading to a simplified experience when using the `sktime` interface.
* the `darts` `LightGBM` model is a basic reduction model, like in `make_reduction`. We should indeed make sure our reduction models work with categorical data out of the box.

My advice would be, after or in parallel to enabling framework support for categorical data, I would start with `skforecast`, as the internal logic here for an interface is simpler than a potential upgrade of the reduction forecaster.

Then, we could try to upgrade `YfromX` - a simpler version of `make_reduction` - to work with `LightGBM`.

**Comment by yarnabrina** [comment]
@fkiraly @Abhay-Lejith Please take a look at the following MRE. `skforecast` do support categorical features natively without encoders if the model supports as well.

```python
import numpy
import pandas
import catboost
from skforecast.ForecasterAutoreg import ForecasterAutoreg

random_generator = numpy.random.default_rng(seed=0)

sample_data = {
    "target_column": random_generator.standard_normal(size=100),
    "categorical_feature": random_generator.choice(["A", "B", "C"], size=100, replace=True),
    "numeric_feature": random_generator.standard_exponential(size=100),
}
sample_dataset = pandas.DataFrame(data=sample_data)

regressor = catboost.CatBoostRegressor(cat_features=["categorical_feature"])
forecaster = ForecasterAutoreg(regressor, 1)

forecaster.fit(
    sample_dataset["target_column"], exog=sample_dataset[["categorical_feature", "numeric_feature"]]
)  # will work successfully
print(forecaster.fitted)  # will print True
```

**Comment by fkiraly** [comment]
Do you mean MRE or MVP? MRE is typically used in the context of bugs, if it is the acronym I remember correctly, "minimal reproducible example". MVP is "minimal viable product". Just asking because I'm not sure whether you are flagging a bug, or a minimal viable code sample to wrap as a functioning sktime estimator. Which of the two?

**Comment by fkiraly** [comment]
Probably MVP, right? I note the code is working. It is also important imo that it disproves the claim that always some encoding transformer needs to be passed, which seems to be heavily implied by skforecast docs. FYI @abhay-lejith.

**Comment by yarnabrina** [comment]
I meant as a minimal reproducible example to demonstrate that skforecast does not necessarily need specification of categorical encoders by users.

(I did not know that MRE term implies bugs, hope the above clarifies my intention.)

**Comment by fkiraly** [comment]
> I did not know that MRE term implies bugs

Interesting point, there's nothing in the term that would imply it, it's used in that context mostly though.

**Comment by Abhay-Lejith** [comment]
https://github.com/sktime/sktime/issues/6109#issuecomment-2141367213

I tried a few different scenarios, the above example where we directly pass the `object` dtype values works only in the case of catboost.
xgboost and lightgbm both fail with error message stating that `object` dtype is not accepted.
Coverting to `CategoricalDtype` and then passing also does not work as `skforecast` blocks it. The dtype of data within the `CategoricalDtype` cannot be `object`. It must be int or float. So encoding is required in these two cases.

**Comment by Abhay-Lejith** [comment]
@fkiraly @yarnabrina , wanted some inputs on this question I had mentioned in the hackmd proposal. xgboost, lightgbm and catboost are some models that have native support for categorical. But method of passing categorical features varies quite a lot between them (each model has different requirements of dtype, parameter specification etc). So modifying the reduction forecaster and making a general interface to accommodate these might not be a good idea.
So I think it might be better to interface them separately on top of the reduction forecasters? (similar to darts lightgbm, see https://unit8co.github.io/darts/generated_api/darts.models.forecasting.lgbm.html )

**Comment by yarnabrina** [comment]
If I am not wrong, this is very similar to your "design 1" where we add `categorical_features` to specific forecasters. But does it not imply to have separate dedicated estimators for every such model? And, e.g. I do not use time series classification/regression/etc. (@fkiraly please help here), but if those use XGBoost, LightGBM, CatBoost etc. we have to create one of the models three times for forecasting, classification, regression, etc. That seems to be not a scalable design.

As of now, my personal opinion is that if @fnhirwa finishes the Darts adapter work (#5043 being continued in #6454), we will probably have the Darts LightGBM capability directly (or possibly through #5997) for forecasting and creating new specific wrappers for these special models is probably not needed. We should try to be generic to add the support for any model, if possible.

(That reminds me, if your design planning have any assumptions regarding scope of work, e.g. only forecasters or only exogenous features etc. it's better be explicitly mentioned.)

**Comment by Abhay-Lejith** [comment]
> But does it not imply to have separate dedicated estimators for every such model?

I meant only for these reduction cases. Because otherwise, I think in reduction classes we will have to do something like 
```
if model used is xgboost:
    xgboost specific categorical specification
else if model used is catboost:
    catboost specific cat specification 
else if ..... and so on
```

> we have to create one of the models three times for forecasting, classification, regression, etc. That seems to be not a scalable design.

I am not familiar with time series classification, I will look into it.

> That reminds me, if your design planning have any assumptions regarding scope of work, e.g. only forecasters or only exogenous features etc. it's better be explicitly mentioned.

Yes, I will iterate through the proposal once more based on @fkiraly 's  feedback and make it more concrete with information like this.

**Comment by fkiraly** [comment]
Re the above, I agree with @yarnabrina that we should not have 10 models for 10 estimators where it could be a much smaller number if we use composition smartly - especially since we may be looking at multiplicative scaling across learning tasks.

In the cleanest of all designs, the different tabular estimators are themselves aware of which features are categorical, or they follow a unified design for the user to specify, but perhaps we cannot expect this.

The if/else handling would indeed be ugly as it is hard-coded, and not extensible, e.g., if we add one more estimator. However, we could go for a hybrid design where:

* we hard-code configurations smartly for frequently used components
* we think carefully about optional arguments by which we could let the user add this configuration. By default it is set to `"auto"` with some smart rules for inferring the config, but the user can pass a dict or function or similar.

This sounds a bit abstract, but your thinking around the classes is indeed spot-on in localizing the challenge, @Abhay-Lejith! This is what I meant with, let's try to be very concrete with examples, so it has now flagged up that in a naive design we will have to hard-code the translation between "which variables are categorical" from inspection of the frame to the 3rd party model specification, per model. To repeat, this is indeed sth I consider a major challenge, and "finding" it shows that your design thought process works! Let's try to write the challenge down with a few examples.

**Comment by Abhay-Lejith** [comment]
@fkiraly @yarnabrina These are the next work items I see in line with the agreed design:
1. add tag to forecasters for categorical capability
    - something like `capability:categorical_in_X`. Set default to False.

2. remove the currently present checks for categorical dtype which fail immediately.
3. feature_kind PR is almost complete, only adding few additional tests remain.

4. a function to perform the internal default encoding logic.
    - in `datatypes/_convert_utils` we can create an `_encode.py` and keep the encoding logic here.
    - import this and call the encode function in `_check_X_y` in `forecasting/base/_base.py` 
    - in _check_X_y we will do something like this:
    ```python
    if not self.get_tag("capability:categorical_in_X"):
        # call encode function
    ```
    - This encoding will have to be done before the `_check_is_scitype` call in `check_X_y` . This is because encoding will create new columns and thus can change the mtype of the input X (Eg. from a single column pd.Series to a pd.DataFrame). 

5. while I was working on the LabelEncoder fix, I had also noted then that OneHotEncoder did not work in sktime pipeline. This compatibility needs to be fixed too.
6. check and ensure all encoders work within the `ColumnEnsembleTransformer` as well.

**Comment by Abhay-Lejith** [comment]
@fkiraly @yarnabrina Here is the sequence in which I think PRs need to be managed:

1. Feature_kind pr - done, can be merged #6490 
2. Adding categorical capability tag pr - pr needs to be created, can be merged once done

3. Remove currently present checks for categorical which fail immediately - old pr exists, #5886 
4. Raise error in yes/no case as initial step - pr needs to be created
3 and 4 need to be merged together. I think both can be done in same pr.

5. replace error raised in yes/no case with default internal encoding logic . - pr needs to be created, can be merged once done.

At this stage, categorical support in forecasters independently should be ready.

6. Think about reduction forecasters and similar cases where user can specify forecaster/regressor to use. Here, categorical support will depend on what is user specified.

Next will be pipelines and similar.

7. `OneHotEncoder` compatibility with pipeline fix.
8. Check if encoders work in `ColumnEnsembleTransformer` and create pr to fix if necessary.

9. Documentation

**Comment by yarnabrina** [comment]
This makes sense to me. I'm not sure if it matches with @fkiraly 's plans, but my plan is not consider these in current cycle (0.31) and do it in July end one (0.32). This is because current PR's if merged will lead to very less capability to end users in absence of estimators' support.

**Comment by fkiraly** [comment]
Ah, sorry, @Abhay-Lejith, I should also reply here, not just in discord.

Yes, agree that this is a very sensible and detailed plan! Some details needed after step 6 and further, but I agree with the general sequence and direcation of rework!

The slicing of 4/5 is based on discussion in our meeting, it introduces another slicing to make the stepwise strategy easier: first we turn off checks and introduce categorical tags, and raise an exception if the user passes categorical, in the yes/no case. Only after that, in a separate step, we add default encoding.

This should isolate the location where the default encoding needs to be added, before transplanting a major piece of logic there, separating "isolation" and "addition", and making the gradual implementation strategy easier.

So, indeed, agreed with the plan.

**Comment by Abhay-Lejith** [comment]
FYI @yarnabrina, summarizing discussion @fkiraly and I had regarding the question I raised on the step of adding encoding logic. Please let me know your thoughts on it as well.
Here is the issue:
In preparation for making the encoding logic, I was experimenting with the encoders and tabular to series adaptor. There are some problems and there are two broad ways of going about it.
![Screenshot from 2024-07-15 13-50-22](https://github.com/user-attachments/assets/e02a5f92-40ed-4608-9696-54dc3245adb2)

<details>
<summary> Discussion </summary>
<br>
- FK agrees with these being the two main options <br>
- would suggest to use OHE directly <br>
        - enabling TTS adapter support is a separate issue <br>
        - the "left" alternative would requier solving both problems <br>
    - also potential issue: how to deal with multiple mtypes? <br>
        - FK: TTS would deal with this <br>
        - AL: played around, TTS applies it instance-wise <br>
        - FK: I made a PR recently to enable pooling <br>
            - https://github.com/sktime/sktime/pull/6752 <br>
        - AL: btw sklearn removes index, but applies pooled globally <br>
        - AL: we could also do conversion to/from mtype where we can deal with encoding, and then convert back <br>
            - FK: was thinking about this and feel uneasy, but agree this may be the only "general" solution <br>
        - AL: could use 6752,but that would mean left branch <br>
            - FK: so we need to check whether 6752 solves the "OHE cannot be used" problem, I would guess it does not <br>
</details>

**Comment by Abhay-Lejith** [comment]
@fkiraly @yarnabrina Update: I was looking at #6752 by FK which adds the global pooling for TTS adaptor and that works correctly. However the OneHotEncoder is still incompatible. I decided to take a look at the error anyway in case it was nothing too big. Turns out it is simply because the sklearn encoder returns a sparse matrix by default and in the adaptor we initialize the output df with this which causes error. So setting `sparse_output=False` in the encoder solves the issue. 
Since its a simple fix, the left branch seems more viable now as problems 2. and 3. are taken care of. We just have to deal with problem 1. This can be done by extending the parent_type of the cat capability tag to transformers and setting it to `True` for the encoders. 
This way we can use the adaptor for the encoding logic instead of attempting to write it from scratch using only the sklearn encoder dealing with the mtype conversions and other problems.

**Comment by yarnabrina** [comment]
I personally prefer left branch anyways, so if #6752 helps it, great 🥳 

Also just to confirm, we are not fixing the encoder choice to OHE, and it was just an example here, right?

---

Regarding sparsity, probably a long term question, how do you plan to address if user create the entire pipeline including encoding and try to use sparse outputs? If there are too many levels of the categorical features, that helps in most cases wrt performance?

**Comment by Abhay-Lejith** [comment]
> Also just to confirm, we are not fixing the encoder choice to OHE, and it was just an example here, right?

For the internal encoding, we decided OHE as the fixed default. If the user wants a different encoding option, they have to pipeline it themselves. This was what we decided yes?

> Regarding sparsity, probably a long term question, how do you plan to address if user create the entire pipeline including encoding and try to use sparse outputs? If there are too many levels of the categorical features, that helps in most cases wrt performance?

For enabling sparse outputs, we will have to modify the adaptor logic accordingly. Maybe it is worth looking into it now itself, instead of making a fix to set sparse_output to False and bypass the error for now. I'll have a look.

**Comment by Abhay-Lejith** [comment]
Seeing how left branch makes more sense now, these are the next steps I propose to implement step 5 of https://github.com/sktime/sktime/issues/6109#issuecomment-2200492018 i.e. to replace error with encoding logic in y/n case.

1. Extend tag to transformers. Set default to False in base class. Set to True in transformers which can take categorical X natively.  This will include the encoders obviously and we can also do it for other applicable transformers like `ColumnSelect` for example.
2. #6752 needs to merged enabling global pooling.
3. TTS fixes to make onehot encoder compatible.
4. Write internal encoding logic for forecasters using the TTS adaptor which now supports categorical and one-hot because of 1,2 and 3.

**Comment by fkiraly** [comment]
Hm, I will need some time to catch up here. Shall we first merge some pull requests to get `main` closer to the development horizon?
Re sparse output, that is indeed a good question, whether supporting that would be useful, though it might create a detour as it would be a new data type.

**Comment by yarnabrina** [comment]
I think it's better to consider support of sparse outputs as a stretch goal for time being, but I think we do need to consider that sooner or later. For example in the pipeline case, users can use OHE with sparse being true (e.g. if users want to tune encoding method between LE, OE, and OHE) and we need to be ready to handle that (either by supporting or by failing with proper message).

Regarding PR sequence, assuming test-all passes in @Abhay-Lejith 's fork, is this the planned merge sequence for now?

#6704 -> #6490 -> #6732 

I'm happy with #6704 and #6732, but I didn't go through #6490 in details. @fkiraly do you have any concerns for any of these 3 PR's?

**Comment by Abhay-Lejith** [comment]
Sequence looks good, #5886 is not a must though because I have already picked commits from it and added in #6732 .

**Comment by Abhay-Lejith** [comment]
> Extend tag to transformers. Set default to False in base class. Set to True in transformers which can take categorical X natively. This will include the encoders obviously and we can also do it for other applicable transformers like ColumnSelect for example.

- re this step
    - for setting categorical support tag in TTS adapter 
        - depends on sklearn transformer passed
        - different ways to deal with this:
            - if checks in init like below
                - ```
                    supported_list=[labelencoder, onehotencoder, ordinalencoder]
                    if transformer is in supported_list:
                        set cat support tag to True
                    ```
        - using isinstance will only cover onehot and ordinal as both are instances of `_BaseEncoder`. But `LabelEncoder` does not come under this. There is no other common inherited class to distinguish the encoders using isinstance.
        - are there any other alternatives? @fkiraly @yarnabrina

**Comment by yarnabrina** [comment]
I don't have an alternative suggestion readily, but I think we definitely have to allow beyond just `scikit-learn` estimators. For example, this is a quite common library:

https://github.com/scikit-learn-contrib/category_encoders

It has a `BaseEncoder` object, which may help us. That needs to be explored though, I haven't gone in too details. If it is enough, we need to decide whether we should add a new soft-dependency, or will it become a core dependency.

**Comment by fkiraly** [comment]
From discussion in the meeting today:

I think we need to ensure we have working examples of estimators, and test coverage, for the case of the "categorical support" tag being True.

The problem that @Abhay-Lejith flagged also needs to be addressed explicitly - there are estimators where it seems impossible to tell whether a given parameter combination supports categoricals or not - for instance, the `TabularToSeriesAdapter`, since that property is *not inspectable* in `sklearn` (according to @Abhay-Lejith).

Given this, we need to address this situation. One way would be to do this consistently with `sklearn`: they let categoricals through, and the estimator may crash or not. Translated to the tag system, it would mean setting the tag to "True" if there is at least one parameter combination that could support categoricals.

**Comment by yarnabrina** [comment]
Mentioning a third approach for both of your consideration: allow categorical support for for transformers that are for categorical encoding and that purpose only -> restrict to [BaseEncoder](https://github.com/scikit-learn-contrib/category_encoders/blob/11fbba6520341e9b960d35dafd44704a67b5bafe/category_encoders/utils.py#L225)

Obvious cons:

1. Need external dependency (but popular and very less additional dependencies: https://github.com/scikit-learn-contrib/category_encoders/blob/11fbba6520341e9b960d35dafd44704a67b5bafe/setup.py#L34)
2. Need to know whether these cover all three of sklearn encoders -> whether they will pass the check (unlikely) or failing we can map to another estimator for inheritance (I'd like to avoid any explicit mapping, if possible)
3. (Please add)

---

I don't think we should code in a way that restricts to only 3 possible encoders in sklearn, so I also feel the same way about extensibility as @fkiraly mentioned above. Admittedly, 
the above possibly adds a similar restriction of encoders being from this external package only, but with more options and an extensible base class.

---

> I think we need to ensure we have working examples of estimators, and test coverage, for the case of the "categorical support" tag being True, before doing anything else.
> 
> Why: without this, we do not test one of two critical cases of the recent extension. Users will see the tag and use it in extension, but may then face bugs or outright failures, because we have failed to cover the case with tests and examples.

I agree 100%

**Comment by yarnabrina** [comment]
As far as I know, this package contain only estimators that encode categorical variables in different ways. So, the check is possibly not required at all, as all of them supports categorical values.

**Comment by fkiraly** [comment]
that would not solve the general problem though

**Comment by Abhay-Lejith** [comment]
@fkiraly @yarnabrina I opened an issue on sklearn to ask about the tags and they do have tags which are used by them in testing but can also be used for inspection via the `_get_tags()` method. It returns a dictionary and the field `X_types` takes values such as '2darray', 'string', 'sparse', 'categorical', dict, '1dlabels' and '2dlabels' . We should be able to use the categorical value to identify the encoders at least. 
But I have noticed that the `LabelEncoder` only has the value '1dlabels' and not 'categorical' since it is used only for the target `y`. The alternative `OrdinalEncoder` which does the same thing but for `X` can be used instead.

**Comment by yarnabrina** [comment]
I don't think we can know whether users will use label encoder in pipelines or not, so how will we force them to use ordinal encoder?

Also, I think we should not restrict ourselves to just 3 encoders from sklearn. As @fkiraly also mentioned above, we need to have a way to identify ANY categorical encoders that are sklearn compatible. Please correct me if I misunderstood. 

(@Abhay-Lejith can you please link the issue you created on sklearn?)

**Comment by fkiraly** [comment]
@abhay-lejith, can you kindly link the issue?

Regarding sklearn tags: it should work if we do "categorical *or* "1dlabels *or* 2dlabels", right?

For a test case of introducing categorical support, I think we have two good choices:

* `TabularToSeriesAdaptor` - as discussed, a transformer
* `YfromX` - a very basic reduction forecaster

**Comment by Abhay-Lejith** [comment]
Here's the issue : https://github.com/scikit-learn/scikit-learn/issues/29539
And here's the sklearn docs about the tag - https://scikit-learn.org/stable/developers/develop.html#estimator-tags

> For a test case of introducing categorical support, I think we have two good choices:
> 
> * `TabularToSeriesAdaptor` - as discussed, a transformer
> * `YfromX` - a very basic reduction forecaster

Re this, should we do the forecaster example first before moving to extending to transformers.

**Comment by fkiraly** [comment]
> Re this, should we do the forecaster example first before moving to extending to transformers.

I do not have a strong opinion about sequencing - I would start with the case where the code and logic is more mature already, by your judgment.

**Comment by Abhay-Lejith** [comment]
> I do not have a strong opinion about sequencing - I would start with the case where the code and logic is more mature already, by your judgment.

Hm, I'm a bit confused. I thought the issue was that you didn't want me to move to adding support for transformers before we had at least one example in forecasting where the tag is set to True. Or is the test using the dummy forecaster sufficient?

**Comment by fkiraly** [comment]
> Hm, I'm a bit confused. I thought the issue was that you didn't want me to move to adding support for transformers before we had at least one example in forecasting where the tag is set to True.

The "do not do X before Y" statement I am trying to make is a bit different. X = "automatic encoding", not "tag in transformers". The Y is "we know the tag works in both settings of forecasters and transformers".

> Or is the test using the dummy forecaster sufficient?

No, because that one ignores `X`, and the tag is about `X`.

> In my opinion, I think that test is sufficient to ensure that the tag works correctly by letting through categorical values without raising error.

I still think we should have at least one, better two, non-trivial examples of estimators working before moving on. So, optimally: two forecasters, two transformers.

**Comment by Abhay-Lejith** [comment]
> No, because that one ignores X, and the tag is about X.
By the way, I am also setting the ignores X tag to false in the test. So the test ensures that the categorical data passes through the whole boilerplate and does not throw any error.

**Comment by fkiraly** [comment]
yes, it's a good test and we should keep it!

It does not replace a "real" estimator with those properties though.

**Comment by Abhay-Lejith** [comment]
@fkiraly @yarnabrina I've been thinking about how we can deal with the tag setting for these compositor type forecasters like the recently added `skforecast` adapter for reduction where categorical support will depend on the model used in reduction. 
The main thing here is that the regressors which support categorical natively are mostly from other libraries like `xgboost`, `catboost` and `lightgbm`. So there isn't really a way as far as I can see to check whether the regressor used in reduction can support natively because these aren't from `sktime/sklearn`. 
What I think is best is that we set the support tag to `True` for these reduction compositors so that users are not completely blocked off from using categorical. And keep the expectation that the user will use categorical data with regressors which can handle it. We can also add a list in documentation of regressors which can be used for categorical data. Worst case is that user uses a regressor which can't handle categorical data and passes it cat data, this will just lead to the error coming from the regressor itself, which is not an issue I feel. 
Please let me know what you think asap. If we can get this figured out then I can move to implementing the encoding logic for forecasters as the transformers extension is done.

**Comment by fkiraly** [comment]
That is perhaps the easiest way to address this, I agree - see my comment above: https://github.com/sktime/sktime/issues/6109#issuecomment-2243105921

I would in-principle be fine with this.

Question: as we discovered above, we can actually inspect the property in `sklearn` and set the tag dynamically, accordingly. Why is this not your preference?

**Comment by Abhay-Lejith** [comment]
> Question: as we discovered above, we can actually inspect the property in sklearn and set the tag dynamically, accordingly. Why is this not your preference?
Yes this works for the sklearn transformers and it is what I have done in the pr for transformers extension here: https://github.com/sktime/sktime/pull/6869

The above comment was related to the regressors used in reduction which are mostly not from sklearn.

**Comment by fkiraly** [comment]
oh, I see - like xgboost, which is sklearn compatible but do I understand you right, it does not follow the tag structure and conventions of sklearn itself?

**Comment by Abhay-Lejith** [comment]
> Do I understand you right, it does not follow the tag structure and conventions of sklearn itself?

Yes, precisely. Other examples are catboost and lightgbm.

**Comment by fkiraly** [comment]
I see. In that case, the only solution that I can see and which does not require us to maintain a "parallel tag database" is the one you propose.
